harness/determined/pytorch/_pytorch_context.py

```diff
@@ -57,4 +82,224 @@ def get_lr_scheduler(self) -> Optional[pytorch.LRScheduler]:
+        # horovod.DistributedOptimizer because torch.optim._LRScheduler directly access the
+        # optimizer parameter groups, which are also Horovod.DistributedOptimizer relies on.
+        wrapped = pytorch.LRScheduler(lr_scheduler, step_mode)
+        self.lr_schedulers.append(wrapped)
+        return wrapped
+
+    def _init_device(self) -> None:
+        self.n_gpus = len(self.env.container_gpus)
+        if self.hvd_config.use:
+            check.gt(self.n_gpus, 0)
+            # We launch a horovod process per GPU. Each process
+            # needs to bind to a unique GPU.
+            self.device = torch.device(hvd.local_rank())
+            torch.cuda.set_device(self.device)
+        elif self.n_gpus > 0:
+            self.device = torch.device("cuda", 0)
+        else:
+            self.device = torch.device("cpu")
+        check.is_not_none(self.device)
+
+    def _to_device(self, data: pytorch._Data) -> pytorch.TorchData:
+        return pytorch.to_device(
+            data, self.device, self.warning_logged[_WarningLogs.FAILED_MOVING_TO_DEVICE]
+        )
+
+    def _use_amp(self) -> bool:
+        return self.mixed_precision != "O0"
+
+    def _configure_amp(
+        self,
+        model: torch.nn.Module,
+        optimizers: Union[torch.optim.Optimizer, List[torch.optim.Optimizer]],  # type: ignore
+        mixed_precision: str = "O0",
+    ) -> Tuple[torch.nn.Module, List[torch.optim.Optimizer]]:  # type: ignore
+        self.mixed_precision = mixed_precision
+        if mixed_precision != "O0":
+            if self.hvd_config.use:
+                check.eq(
+                    self.hvd_config.aggregation_frequency,
+                    1,
+                    "Mixed precision training (AMP) is not supported with "
+                    "aggregation frequency > 1.",
+                )
+
+            check.true(
+                torch.cuda.is_available(),
+                "Mixed precision training (AMP) is supported only on GPU slots.",
+            )
+            check.false(
+                not self.hvd_config.use and self.n_gpus > 1,
+                "To enable mixed precision training (AMP) for parallel training, "
+                'please set `resources["optimized_parallel"] = True`.',
+            )
+
+            logging.info(
+                f"Enabling mixed precision training with opt_level: {self.mixed_precision}."
+            )
+            model, optimizers = apex.amp.initialize(
+                model,
+                optimizers,
+                opt_level=self.mixed_precision,
+                verbosity=1
+                if self.distributed.get_rank() == 0 or self.env.experiment_config.debug_enabled()
+                else 0,
+            )
+            self.model = model
+            if not isinstance(optimizers, list):
+                self.optimizers = [optimizers]
+        return model, optimizers
+
+    def _should_communicate_and_update(self) -> bool:
+        if self.current_batch_idx is None:
```

Comment:
non-blocking: could we make getters and setters for `self.current_batch_idx`

Reply:
We don't provide functions to set `current_batch_idx` to users and it's used only in one place. Sound unnecessary to have getters and setters